File: Permutations of a given string - GFG/permutations-of-a-given-string.java
//{ Driver Code Starts
import java.util.*;
import java.lang.*;
import java.io.*;
class GFG
{
	public static void main(String[] args) throws IOException
	{
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int t = Integer.parseInt(br.readLine().trim());
        while(t-->0)
        {
            String S = br.readLine().trim();
            Solution obj = new Solution();
            List<String> ans = obj.find_permutation(S);
            for( int i = 0; i < ans.size(); i++)
            {
                System.out.print(ans.get(i)+" ");
            }
            System.out.println();
                        
        }
	}
}


// } Driver Code Ends

class Solution {

         public void find(int index,String s,List<String>ans)

         {

               if(index==s.length())

               {

                     if(!ans.contains(s))

                     ans.add(s);

                     return;

               }

               for(int i=index;i<s.length();i++){

                    s=swap(i,index,s);

                    find(index+1,s,ans);

                    s=swap(i,index,s);

               }

         }

        public String swap(int i,int j,String s)

        {

               char [] ch=s.toCharArray();

                char c=ch[i];

                ch[i]=ch[j];

                ch[j]=c;

                return String.valueOf(ch);

        }

    public List<String> find_permutation(String S) {

        // Code here

        List<String>ans=new ArrayList<>();

        find(0,S,ans);

        Collections.sort(ans);

        return ans;

    }

}


































    

File: Permutations of a given string - GFG/permutations-of-a-given-string.py
#User function Template for python3

class Solution:
    def solve(self, ans, nums, i):
        if i == len(nums):
            if "".join(nums[:]) not in ans:
                ans.append("".join(nums[:]))
            return
        for j in range(i, len(nums)):
            nums[i], nums[j] = nums[j], nums[i]
            self.solve(ans, nums, i+1)
            nums[i], nums[j] = nums[j], nums[i]
    
    def find_permutation(self, S):
        ans = []
        nums=list(S)
        self.solve(ans, nums, 0)
        ans=sorted(ans)
        return ans


#{ 
 # Driver Code Starts
#Initial Template for Python 3


if __name__ == '__main__':
	t=int(input())
	for i in range(t):
		S=input()
		ob = Solution()
		ans = ob.find_permutation(S)
		for i in ans:
			print(i,end=" ")
		print()
# } Driver Code Ends